DRILL-6115: SingleMergeExchange is not scaling up when many minor fra…

HanumathRao opened the pull request:
…gments are allocated for a query.

Currently a singlemerge exchange is merging all the fragment streams on foreman. This can cause cpu bottleneck and also huge memory consumption at the foreman. 

This PR contains changes to introduce a new Multiplex Operator called OrderedMuxExchange which merges the minor fragment streams pertaining to one drillbit and send as one output stream to the foreman. 

The existing multiplex mechanism is used to introduce these operators.

Please review this PR.

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/visitor/PrelVisitor.java:
```diff
@@ -35,7 +38,9 @@
   public RETURN visitJoin(JoinPrel prel, EXTRA value) throws EXCEP;
   public RETURN visitProject(ProjectPrel prel, EXTRA value) throws EXCEP;
-
+  public RETURN visitHashToRandomExchange(HashToRandomExchangePrel prel, EXTRA value) throws EXCEP;
```
Are 3 new methods necessary? Can `visitExchange` delegate to `prel` or use instance of?

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/base/PhysicalVisitor.java:
```diff
@@ -55,10 +56,10 @@
 
 
   public RETURN visitExchange(Exchange exchange, EXTRA value) throws EXCEP;
+  public RETURN visitSingleMergeExchange(SingleMergeExchange exchange, EXTRA value) throws EXCEP;
```
The same question as for `PrelVisitor.java`. Is it necessary to have separate `visitSingleMergeExchange`?

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/config/OrderedMuxExchange.java:
```diff
@@ -0,0 +1,58 @@
+public class OrderedMuxExchange extends AbstractMuxExchange {
+  private final List<Ordering> orderings;
+
+  public OrderedMuxExchange(@JsonProperty("child") PhysicalOperator child, List<Ordering> orderings) {
```
Json annotation for orderings?

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/config/OrderedMuxExchange.java:
```diff
@@ -0,0 +1,58 @@
+    createSenderReceiverMapping();
+
+    List<MinorFragmentEndpoint> senders = receiverToSenderMapping.get(minorFragmentId);
+    if (senders == null || senders.size() <= 0) {
```
Add debug level info for `receiverToSenderMapping` and minorFragmentId.

vrozov commented on exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/xsort/TestSortSpillWithException.java:
```diff
@@ -59,6 +59,7 @@ public static void setup() throws Exception {
     ClusterFixtureBuilder builder = ClusterFixture.builder(dirTestWatcher)
         .configProperty(ExecConstants.EXTERNAL_SORT_SPILL_THRESHOLD, 1) // Unmanaged
         .configProperty(ExecConstants.EXTERNAL_SORT_SPILL_GROUP_SIZE, 1) // Unmanaged
+        .configProperty(ExecConstants.EXTERNAL_SORT_MAX_MEMORY, 10 * 1024 * 1024) //use less memory for sorting.
```
Why is the change necessary?

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/visitor/InsertLocalExchangeVisitor.java:
```diff
@@ -20,133 +20,34 @@
 
     if (isMuxEnabled || isDeMuxEnabled) {
-      return prel.accept(new InsertLocalExchangeVisitor(isMuxEnabled, isDeMuxEnabled), null);
+      return prel.accept(new InsertLocalExchangeVisitor(options), null);
```
Since the local variables isMuxEnabled/disabled are not being used anymore, you can remove them on lines 33, 34.

HanumathRao commented:
@vrozov  Thank you for reviewing the code. I have incorporated all the review comments. Please let me know if anything needs to be changed.

amansinha100 commented on exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/TestOrderedMuxExchange.java:
```diff
@@ -0,0 +1,116 @@
+    try (ClusterFixture cluster = builder.build();
+         ClientFixture client = cluster.clientFixture()) {
+      client.alterSession(ExecConstants.SLICE_TARGET, 10);
+      String sql = "SELECT id_i, name_s10 FROM `mock`.`employees_10K` ORDER BY id_i limit 10";
```
I am not sure how the table is organized..does it have already ordered id_i column ? if so, we should use a different column.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/config/OrderedMuxExchange.java:
```diff
@@ -0,0 +1,64 @@
+      logger.debug(String.format("Minor fragment %d, receives data from following senders %s", minorFragmentId, senders));
+    }
+
+    return new MergingReceiverPOP(senderMajorFragmentId, senders, orderings, false);
```
The HashToMergeExchange creates a MergingReciver with spooling TRUE, whereas the SingleMergeExchange creates one with spooling FALSE.  Although we don't test the spooling, I feel the new OrderedMuxExchange should probably have the same spooling setting as the HashToMergeExchange since both do the merge on local drill bits vs the foreman.

amansinha100 commented on exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/TestOrderedMuxExchange.java:
```diff
@@ -0,0 +1,116 @@
+
+      String explainText = client.queryBuilder().sql(sql).explainText();
+      assertTrue(explainText.contains(ORDERED_MUX_EXCHANGE));
+      validateResults(client.allocator(), client.queryBuilder().sql(sql).results());
```
I think a simpler way to validate results would be to compare with a plan that does not have this mux exchange; that will allow arbitrary data types to be handled since currently you have a hardcoded IntVector comparison which limits the types of queries one can test.

amansinha100 commented:
@HanumathRao I have a few comments in the JIRA for the overall design; we can discuss.

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/config/OrderedMuxExchange.java:
```diff
@@ -0,0 +1,64 @@
+
+  @Override
+  public Receiver getReceiver(int minorFragmentId) {
+    createSenderReceiverMapping();
```
Consider moving this functionality to the parent class and keep only creating an instance of concrete `MergingReceiver` in the subclass.

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/ExchangePrel.java:
```diff
@@ -34,4 +37,14 @@ public ExchangePrel(RelOptCluster cluster, RelTraitSet traits, RelNode child) {
+   * @param child input to the new muxPrel or new Exchange node.
+   * @param options options manager to check if mux is enabled.
+   */
+  public Prel getMuxPrel(Prel child, OptionManager options) {
```
Consider renaming to `constructMuxPrel`.

HanumathRao commented:
@amansinha100 @vrozov  Thank you for the review. I have addressed all the review comments. Please let me know if any changes are required.

The commits are organized such that one commit is for refactoring the existing code and the second one is specific to the changes required for this JIRA. This is done for ease of reviewing the code.

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/config/AbstractMuxExchange.java:
```diff
@@ -90,6 +92,24 @@ public Sender getSender(int minorFragmentId, PhysicalOperator child) {
+
+    List<MinorFragmentEndpoint> senders = receiverToSenderMapping.get(minorFragmentId);
+
+    logger.debug(String.format("Minor fragment %d, receives data from following senders %s", minorFragmentId, senders));
```
Use SLF4J smart logging instead of `String.format`.

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/config/UnorderedMuxExchange.java:
```diff
@@ -37,14 +37,8 @@ public UnorderedMuxExchange(@JsonProperty("child") PhysicalOperator child) {
-
+  protected Receiver getReceiverInternal(int oppositeMajorFragmentId,
+                                         List<MinorFragmentEndpoint> senders, boolean spooling) {
     return new UnorderedReceiver(senderMajorFragmentId, senders, false);
```
`spooling` parameter is ignored, is this expected?

vrozov replied:
Consider creating helper method like `getSenders()` that will return ` List<MinorFragmentEndpoint>` instead of `getReceiverInternal()`.

HanumathRao replied:
Done.

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/visitor/InsertLocalExchangeVisitor.java:
```diff
@@ -20,133 +20,41 @@
-          new DrillSqlOperator(funcName, inputFields.size(), true, false);
-      return rexBuilder.makeCall(op, inputFields);
+  private static boolean isMuxEnabled(OptionManager options) {
+    if (options.getOption(PlannerSettings.MUX_EXCHANGE.getOptionName()).bool_val ||
```
use `return` instead of `if`

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/config/AbstractMuxExchange.java:
```diff
@@ -92,24 +92,19 @@ public Sender getSender(int minorFragmentId, PhysicalOperator child) {
     List<MinorFragmentEndpoint> senders = receiverToSenderMapping.get(minorFragmentId);
 
-    logger.debug(String.format("Minor fragment %d, receives data from following senders %s", minorFragmentId, senders));
+    logger.debug("Minor fragment %d, receives data from following senders %s", minorFragmentId, senders);
```
use {} in place of %d and %s.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/SingleMergeExchangePrel.java:
```diff
@@ -93,6 +94,21 @@ public PhysicalOperator getPhysicalOperator(PhysicalPlanCreator creator) throws
+  @Override
+  public Prel constructMuxPrel(Prel child, OptionManager options) throws RuntimeException {
+    Prel outPrel = child;
+    if (options.getOption(PlannerSettings.ORDERED_MUX_EXCHANGE.getOptionName()).bool_val) {
```
@HanumathRao I think the ordered_mux should be created when both mux and ordered_mux flags are enabled.  Users who disable the global 'mux' flag would very likely expect that all mux exchanges are disabled.

HanumathRao replied:
@amansinha100  Thanks for the review . I have done the needed code changes to fix it. Please let me know if anything is required.

amansinha100 commented:
LGTM.  +1

vrozov commented:
LGTM, please squash two last commits.

HanumathRao commented:
@amansinha100  @vrozov  Thanks for the review. I have squashed all the commits into two commits.

Please merge these two commits individually into the apache master branch. First commit is for refactoring existing code. Second commit is about fixing this JIRA request.
